python/pyarrow/tests/test_sparse_tensor.py

```diff
@@ -434,6 +434,21 @@ def test_sparse_coo_tensor_scipy_roundtrip(dtype_str, arrow_type):
     assert sparse_tensor.has_canonical_format
     assert out_scipy_matrix.has_canonical_format
 
+    scipy_matrix = coo_matrix([[0, 0], [0, 0]])
```

Comment:
Does this actually create a tensor with zero elements?

Reply:
Yeah, coo_matrix automatically ignores zeros, I'll add an assertion checking for number of elements to make this explicit.

Reply:
Well actually it's:
```
    scipy_matrix = coo_matrix(([0], ([0], [0])), shape=(2, 4))
    assert scipy_matrix.nnz == 1
    scipy_matrix = coo_matrix([[0, 0], [0, 0]])
    assert scipy_matrix.nnz == 0
```

Reply:
Right, but the shape is all non-zero.

Reply:
Well those are actually different constructors (shape seems to be ignored in this case):
```
1.
coo_matrix((data, (i, j)), [shape=(M, N)])
to construct from three arrays:
data[:] the entries of the matrix, in any order
i[:] the row indices of the matrix entries
j[:] the column indices of the matrix entries

    scipy_matrix = coo_matrix(([0], ([0], [0])))
    assert scipy_matrix.nnz == 1

2.
coo_matrix(D)
with a dense matrix D

    scipy_matrix = coo_matrix([[0, 0], [0, 0]])
    assert scipy_matrix.nnz == 0
```

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html

Reply:
But this doesn't address the issue, does it?

Reply:
The reported issue was due to `scipy.coo_matrix` returning a sparse matrix with a dimension of zero size for an all zeros dense tensor. This is not an issue if the sparse matrix is created from components. Propose change tests from dense creation path in Python with `scipy.coo_matrix`, `scipy.csr_matrix` and `sparse.COO`. It also tests C++ SparseTensor creation from a dense tensor with one zero-sized dimension for `SparseCOOTensor`, `SparseCSRMatrix`, `SparseCSRMatrix` and `SparseCSFTensor`.

Reply:
I might be missing the point though :D

Reply:
I believe @rok has the correct test here but I could be wrong.  The original issue is about a sparse matrix with a valid non-zero shape but no elements:

```
>>> scipy.sparse.coo_matrix(numpy.zeros((2,4)), dtype=numpy.float32)
<2x4 sparse matrix of type '<class 'numpy.float32'>'
	with 0 stored elements in COOrdinate format>
>>> numpy.zeros((2,4))
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]])
>>> numpy.zeros((2,4)).shape
(2, 4)
```